blog: architecture driven auth part 1 (BFF)

sixhobbits opened the pull request.

nielthiart committed:
DRAFT: Add part one of architecture-driven auth, BFF

nielthiart committed:
Add BFF SPA comparison diagram

nielthiart committed:
Add BFF demo diagram

kmaida commented on astro/src/content/blog/backend-for-frontend-security-architecture.mdx:
```diff
@@ -0,0 +1,511 @@
+    grant_type: 'authorization_code',
+    code: authCode,
+    client_id: process.env.CLIENT_ID,
+    client_secret: process.env.CLIENT_SECRET, // ← Browser can't do this
```
Given that there has been some kerfuffle about ENV variables not being safe from AI, let's just use `clientID` and `clientSecret` as variables to not claim an "opinion" about this in this particular article.

kmaida reviewed (commented):
Comments are pretty high level here, I didn't review for editorial.

nielthiart committed:
Update BFF draft based on high level content feedback

- Reduce focus on NPM and supply chain attacks.
- Clarify backend vs resource server paths in demo.
- Remove potentially controversial reliance on ENV vars.
- Remove direct mention of competitors in series teaser.

Co-authored-by: Kim Maida <kim.maida@gmail.com>

rideam committed:
wip updates

rideam committed:
fixes

rideam committed:
fix capitalisation check

bethh0rn committed:
Edit: Backend-for-Frontend

bethh0rn committed:
Update backend-for-frontend-security-architecture.mdx

bethh0rn committed:
Merge pull request #315 from ritza-co/edit-architecture-driven-auth-bff

Edit: Backend-for-Frontend

bethh0rn committed:
Consistency edit

kmaida commented on astro/src/content/blog/backend-for-frontend-security-architecture.mdx:
```diff
@@ -0,0 +1,511 @@
+
+### Set up the architecture
+
+To follow along, you need Node.js, npm, and Docker installed.
```
```suggestion
To follow along, you need [Node.js](https://nodejs.org), npm, and [Docker](https://www.docker.com/) installed.
```

kmaida commented on astro/src/content/blog/backend-for-frontend-security-architecture.mdx:
````diff
@@ -0,0 +1,511 @@
+  ```
+
+  <Aside type="note">
+    The resource server is a simple API that requires authentication and returns recipes at the endpoint `/api/recipe`.
````
```suggestion
    The resource server is a simple API that requires authentication and returns recipe JSON at the endpoint `/api/recipe`.
```

kmaida commented on astro/src/content/blog/backend-for-frontend-security-architecture.mdx:
```diff
@@ -0,0 +1,511 @@
+
+Since there is no user session, the backend responds with `loggedIn: false`, and the app shows the **Login** button.
+
+The backend's response includes an `httpOnly` cookie, `p`, with a state value, PKCE code verifier, and challenge. This is stored securely in the browser and sent with every request to the backend until the user establishes a session.
```
```suggestion
The backend's response includes an `httpOnly` cookie, `p`, with a state value, PKCE code verifier, and challenge. This is stored in the browser and sent with every request to the backend until the user establishes a session.
```

kmaida reviewed (commented):
I don't think I can approve reviews anymore since I'm no longer a CODEOWNER. However, with these changes, this article looks good to me. Once changes are implemented, someone on the DevRel team can be pinged for approval.

sixhobbits committed:
Update astro/src/content/blog/backend-for-frontend-security-architecture.mdx

Co-authored-by: Kim Maida <yi.mihi@gmail.com>

sixhobbits committed:
Update astro/src/content/blog/backend-for-frontend-security-architecture.mdx

Co-authored-by: Kim Maida <yi.mihi@gmail.com>

sixhobbits committed:
Apply suggestions from code review

Co-authored-by: Kim Maida <yi.mihi@gmail.com>

sixhobbits committed:
Update astro/src/content/blog/backend-for-frontend-security-architecture.mdx

Co-authored-by: Kim Maida <yi.mihi@gmail.com>

bethh0rn committed:
Consistency edit

bethh0rn committed:
Merge pull request #340 from ritza-co/edit-feedback-bff

Consistency edit

nathan-contino commented on astro/src/content/blog/backend-for-frontend-security-architecture.mdx:
```diff
@@ -0,0 +1,516 @@
+- ✅ Handle sensitive data (PII, financial, healthcare).
+- ✅ Build business-critical applications.
+- ✅ Are subject to compliance requirements.
+- ✅ Control the backend.
```
```suggestion
- Handle sensitive data (PII, financial, healthcare).
- Build business-critical applications.
- Are subject to compliance requirements.
- Control the backend.
```
Emoji checkmarks in lists are an LLM cliche we'd like to avoid.

nathan-contino commented on astro/src/content/blog/backend-for-frontend-security-architecture.mdx:
```diff
@@ -0,0 +1,516 @@
+
+- Clone and run the [BFF demo repository](https://github.com/kmaida/auth-architecture).
+- Try [FusionAuth's Hosted Backend](/docs/apis/hosted-backend) for instant BFF security. The Hosted Backend API is available on all FusionAuth plans, including the free Community plan.
+- Plan your migration from less secure patterns.
```
```suggestion
```

This feels redundant when we spent most of the article explaining how to run this yourself.

nathan-contino commented on astro/src/content/blog/backend-for-frontend-security-architecture.mdx:
```diff
@@ -0,0 +1,516 @@
+- Try [FusionAuth's Hosted Backend](/docs/apis/hosted-backend) for instant BFF security. The Hosted Backend API is available on all FusionAuth plans, including the free Community plan.
+- Plan your migration from less secure patterns.
+
+## Next in this series
+
+Part 2 examines the [Token-Mediating Backend (TMB) architecture pattern](/blog/auth-architecture-part2-TMB) - the middle-ground approach that many auth providers use by default. Learn why it's better than storing all tokens in the browser but still falls short of BFF's security guarantees, explore real implementations of TMB, and discover when using TMB is acceptable despite its limitations.
+
+Part 3 covers the [Browser-Based OAuth Client (BBOC) architecture](/blog/browser-based-oauth-client-security-architecture) - the least secure pattern, which the draft specifically recommends against using for sensitive applications. Learn why some providers still default to this pattern and what you absolutely must do if you're stuck with it.
```
It might be better to comment this out right now and reveal these links once we publish parts 2 and 3. It'll take some time to review them (and for the most mindshare, I'm sure we want to spread out these releases).

We definitely need to remove the links at the very least until we can publish all of the parts of this series -- we don't want 404s (or to introduce errors in the link checker part of our build)!

jamesdanielwhitford replied:
Done [980da84](https://github.com/FusionAuth/fusionauth-site/pull/3942/commits/980da84a3b6f6545c6c3a05a782c68521295b1ab). Commented out the "Next in this series" section with a TODO so we can easily restore it when parts 2 and 3 are ready to publish. 

One other note: the `publish_date` in the frontmatter is currently set to October 2025. We'll need to update that before we ship.

nathan-contino commented on astro/src/content/blog/backend-for-frontend-security-architecture.mdx:
```diff
@@ -0,0 +1,516 @@
+htmlTitle: "BFF security architecture: Surviving supply chain attacks | FusionAuth"
+description: "Learn why Backend-for-Frontend (BFF) is the only auth architecture that survives npm supply chain attacks. Part 1 of 3 in our architecture-driven auth series."
+image: /img/blogs/bff/bff-header.png # TODO: replace image
+authors: Kim Maida
```
Let's fix this header image.

Additionally: I don't mind crediting Kim if she wrote this post, it seems like the right thing to do. But since others participated I _think_ we can add them as co-authors, it feels right to acknowledge everyone's contributions.

jamesdanielwhitford replied:
Kept Kim as the sole author, this post was adapted from a video Kim made. 

[184f73e](https://github.com/FusionAuth/fusionauth-site/pull/3942/commits/184f73e11be3a716517f052b1477645ef4cebffc)

jamesdanielwhitford replied:
For the header image, Ritza hasn't worked on any header images in the past, we would be happy to take a look. If you could send a Figma project etc for us to use your templates.

Otherwise I generated a header image for this guide in the mean time. Which you can keep if you are happy with it. [9f9ecaf](https://github.com/FusionAuth/fusionauth-site/pull/3942/commits/9f9ecafcf00e1704616e156e38d01daf164e2f82)

nathan-contino reviewed (changes requested):
4 (very) minor fixes, and we'll be ready to ship! I'd like to push this ASAP since it's been cooking for so long, so the sooner you can address my comments, the sooner we can ship.

I've decided to launch these blog posts separately, so let's remove all of the references to parts 2 and 3 (though feel free to comment them out with a TODO so we can easily restore the links once we publish those).

Let me know if you need my help with a header image. It isn't a big deal, but we of course want something professional.

Changes requested, but LGTM after that. This is great content, I learned quite a lot!

jamesdanielwhitford committed:
Update astro/src/content/blog/backend-for-frontend-security-architecture.mdx

Co-authored-by: nathan contino <patronus@lambdalatitudinarians.org>

jamesdanielwhitford committed:
Update astro/src/content/blog/backend-for-frontend-security-architecture.mdx

Co-authored-by: nathan contino <patronus@lambdalatitudinarians.org>

jamesdanielwhitford committed:
Comment out Next in this series section until parts 2 and 3 are published

jamesdanielwhitford committed:
Add Niel Thiart as co-author and remove image TODO comment

jamesdanielwhitford committed:
Revert to Kim Maida as sole author

jamesdanielwhitford committed:
Add new header image

nathan-contino commented on astro/src/content/blog/backend-for-frontend-security-architecture.mdx:
```diff
@@ -0,0 +1,513 @@
+---
+publish_date: 2025-10-07
```
```suggestion
publish_date: 2026-04-22
```

How about today?

nathan-contino commented on astro/src/content/blog/backend-for-frontend-security-architecture.mdx:
```diff
@@ -0,0 +1,513 @@
+title: "Backend-for-Frontend: The most secure architecture for browser-based apps"
+htmlTitle: "BFF security architecture: Surviving supply chain attacks | FusionAuth"
+description: "Learn why Backend-for-Frontend (BFF) is the only auth architecture that survives npm supply chain attacks. Part 1 of 3 in our architecture-driven auth series."
+image: /img/blogs/bff/bff-security-architecture-header.png
```
This looks fine enough to me, we don't need anything too fancy anyway.

nathan-contino reviewed (commented):
LGTM, with an updated publish date. Let's finally ship this.

sixhobbits committed:
Update astro/src/content/blog/backend-for-frontend-security-architecture.mdx

Co-authored-by: nathan contino <patronus@lambdalatitudinarians.org>

nathan-contino reviewed (approved):
LGTM!
